Add zenodo-data-release.yml workflow.

zaneselvans opened the pull request:
# Overview

- Add a new `zenodo-data-release.yml` GitHub workflow that calls the `zenodo_data_release.py` script.
- Stop running the `zenodo_data_release.py` script in the nightly builds and instead dispatch the new workflow using `curl` and the GitHub API after the build outputs have been successfully distributed to S3.
- Move the `zenodo_data_release.py` script out of `devtools/zenodo` and into `src/pudl/scripts/` so it can be imported and tested and is available on the user's `$PATH`.
- Refactor `retry_request`, `create_bucket_file`, and `sync_directory` to restart the upload of a file from the beginning after encountering an error, and to do retries on common transient error statuses like 502 Bad Gateway.
- Added tests for the retry behavior.
- Disabled uploading of build outputs to S3 and GCS for `workflow_dispatch` builds, since we immediately delete those outputs anyway, the uploads have been quite robust, and there are egress fees associated with sending the data to S3.
- Update to `actions/checkout@v6` while we're in here.
- Cut the nightly build VM back down to 8 vCPUs and 64 GB of RAM since it's working again, and cheaper per run than the bigger machine (and not that much slower).

Closes #4775
Closes #4290 

## Documentation

- [x] Update the [release notes](https://catalystcoop-pudl.readthedocs.io/en/nightly/release_notes.html): reference the PR and related issues.
- [x] Review and update any other aspects of the documentation that might be affected by this PR.

## Testing

- I'm not really sure how to test how the code responds to this kind of intermittent network flakiness. There's an LLM generated test module but I can't evaluate whether it's testing things meaningfully so we should maybe discard it if someone else really knows how to test this kind of thing.
- I've tested behavior locally by running the script and then turning my wifi on and off to simulate a flaky network, and it has been able to do catch errors and do retries successfully.

## Tasks / Testing

- [x] Run the new workflow using `workflow_dispatch` to make sure it works with trivial data.
- [x] Quiet the logging about ignored files (set to `DEBUG`)
- [x] Run the new workflow using `workflow_dispatch` with a full data release's worth of data.
- [x] Try dispatching the workflow locally using `curl` in the same way as we would in the build script.
- [x] Add appropriate `nightly` dispatch of `zenodo-data-release` workflow to build script.
- [x] Add appropriate `stable` dispatch of `zenodo-data-release` workflow to build script.
- [x] Ensure that if the build is NOT successful that we DO NOT do a Zenodo release.
- [x] Re-open or rewind the "file" (`IOStream`) during retry, otherwise we upload zero bytes.
- [x] #4290 
- [x] Review the PR yourself and call out any questions or issues you have.
- [x] Run a `workflow_dispatch` build with a real ETL and see what happens.

zaneselvans commented on .github/workflows/zenodo-data-release.yml:
```diff
@@ -57,16 +62,20 @@ jobs:
           ZENODO_TOKEN_UPLOAD: ${{ secrets.ZENODO_TOKEN_UPLOAD }}
         run: |
-          python devtools/zenodo/zenodo_data_release.py \
+          zenodo_data_release \
```
Switch to using the installed version of the script.

zaneselvans commented on docker/gcp_pudl_etl.sh:
```diff
@@ -216,7 +228,6 @@ WRITE_DATAPACKAGE_SUCCESS=0
 CLEAN_UP_OUTPUTS_SUCCESS=0
 DISTRIBUTION_BUCKET_SUCCESS=0
 DEPLOY_EEL_HOLE_SUCCESS=0
-ZENODO_SUCCESS=0
```
Note that we will no longer get a Zenodo status directly from the builds. But we will get it from the workflow, and it'll also be sent to the `pudl-deployments` channel.

zaneselvans commented on test/unit/scripts/test_zenodo_data_release.py:
I don't really know how to meaningfully test this kind of sporadic network weirdness and our response to it. This is an LLM suggestion, but if you have real ideas that would be great.  I've tested it a bit locally by turning my wifi on and off during the upload, and it has successfully caught an error and recovered.

krivard replied:
The canonical way is to mock all network requests so that the test can supply a specific response code and verify the resulting behavior is what we want. This is what the LLM does below, using [the built-in `monkeypatch` fixture in pytest](https://docs.pytest.org/en/stable/how-to/monkeypatch.html) (the docs at that link include exactly our use case of mocking requests.get). 

In the PUDL codebase we generally prefer `pytest-mock` ([here's an explainer about the differences between monkeypatch, pytest-mock, and unittest.mock](https://github.com/pytest-dev/pytest/issues/4576#issuecomment-449864333)) but in this case I don't think we need any of the eavesdropping features of pytest-mock, so monkeypatch is probably fine.

krivard commented on docker/gcp_pudl_etl.sh:
```diff
@@ -319,12 +332,18 @@ elif [[ "$BUILD_TYPE" == "workflow_dispatch" ]]; then
+            "s3://pudl.catalyst.coop/nightly/" \
+            "${ZENODO_IGNORE_REGEX}" \
+            "publish" 2>&1 | tee -a "$LOGFILE"
+    fi
```
GHA has some concurrency controls; would they help here?

https://docs.github.com/en/actions/how-tos/write-workflows/choose-when-workflows-run/control-workflow-concurrency

zaneselvans replied:
Ahh, great. I'll create two concurrency groups, one for each of the Zenodo environments.

zaneselvans commented on src/pudl/scripts/zenodo_data_release.py:
```diff
@@ -1,29 +1,69 @@
+    520,  # Web server returned an unknown error (proxy)
+    522,  # Connection timed out (proxy)
+    524,  # A timeout occurred (proxy)
+}
```
Removed 429 from the list.

krivard commented on docs/release_notes.rst:
```diff
@@ -21,6 +21,13 @@ New Data Tests & Data Validations
 
+* Increase the robustness of retries when uploading a PUDL data release to Zenodo,
+  catching common transient error status codes, and restarting file uploads from the
+  beginning of the file when an error occurs rather than uploading an empty file.
```
This sentence is ambiguous. Did you mean:

1. we use retries on all three of (uploading, catching, and restarting)
2. increasing robustness of retries, catching, and restarting are 3 separate-but-related improvements
3. catching and restarting are two examples of how we're increasing robustness of retries

for 1, consider:

> Increase the robustness of retries in three areas of Zenodo interactions: uploading a PUDL data release to Zenodo, catching common transient error status codes, and restarting file uploads from the beginning of the file when an error occurs rather than uploading an empty file.

for 2, consider:

> Increase the robustness of retries when uploading a PUDL data release to Zenodo, catch common transient error status codes, and restart file uploads from the beginning of the file when an error occurs rather than uploading an empty file.

for 3, consider:

> Increase the robustness of retries when uploading a PUDL data release to Zenodo, in two ways: catching common transient error status codes, and restarting file uploads from the beginning of the file when an error occurs rather than uploading an empty file.

krivard commented on src/pudl/scripts/zenodo_data_release.py:
```diff
@@ -1,29 +1,67 @@
+associated with the same original PUDL concept DOI.
+
+This module provides a CLI that handles the process of uploading a new version to
+Zenodo, given a prepared directory of artifacts, typically produced by the PUDL builds.
```
```suggestion
Zenodo, given a prepared directory of artifacts typically produced by the PUDL builds.
```

krivard commented on src/pudl/scripts/zenodo_data_release.py:
```diff
@@ -1,29 +1,67 @@
+Implements resilient retries for all requests to recover from transient network issues
+and Zenodo server flakiness. Zero-byte uploads are prevented.
+
+NOTE: PUDL nightly build outputs are NOT suitable a Zenodo data release unless the
```
```suggestion
NOTE: PUDL nightly build outputs are NOT suitable for a Zenodo data release unless the
```

krivard reviewed (approved):
No blocking feedback; just a buncha documentation nits, some confusion on what kind of zenodo workflow we want to trigger on workflow dispatch etls, and an optional learning opportunity sidequest for comparative mocking studies

zaneselvans commented:
* I updated the docstrings / notes to try and be clearer about what I changed and what the previous situation was.
* I rewrote the test module to use `pytest-mock` and to test both retryable and non-retryable status codes, as well as raising the error after `max_retries`
